src/lib/comms/sol-http-common.c

```diff
@@ -123,3 +120,453 @@ sol_http_escape_string(char **escaped, const char *value)
+    return true;
+}
+
+static int
+_get_protocol(const char *uri, struct sol_str_slice *protocol, char **end)
+{
+    struct sol_str_slice p;
+    size_t i;
+
+    p.data = uri;
+    p.len = 0;
+
+    for (i = 0; uri[i] && uri[i] != ':'; i++, p.len++) {
+        if (!isalnum(uri[i]) && uri[i] != '+' &&
+            uri[i] != '-' && uri[i] != '.') {
+            SOL_ERR("Invalid character at the URI's scheme: %c", uri[i]);
+            return -EINVAL;
+        }
+    }
+
+    if (uri[i] != ':') {
+        SOL_ERR("Could not find the \":\" scheme");
+        return -EINVAL;
+    }
+
+    SOL_DBG("URI Protocol: %.*s", SOL_STR_SLICE_PRINT(p));
+    *end = (char *)p.data + p.len + 1;
+
+    if (protocol)
+        *protocol = p;
+
+    return 0;
+}
+
+static int
+_get_server(char *partial_uri, struct sol_str_slice *server, char **next)
+{
+    char *i;
+    struct sol_str_slice s;
+    size_t j;
+
+    for (j = 0; j < 2; j++, partial_uri++) {
+
+        if (!*partial_uri) {
+            SOL_ERR("Expecting a \"/\" character to parse the server. "
+                "Found: NUL");
+            return -EINVAL;
+        }
+
+        if (*partial_uri != '/') {
+            SOL_ERR("Expecting a \"/\" character to parse the server. "
+                "Found: %c", *partial_uri);
+            return -EINVAL;
+        }
+    }
+
+    if (!*partial_uri) {
+        SOL_ERR("Missing host information.");
+        return -EINVAL;
+    }
+
+    s.data = partial_uri;
+
+    for (i = partial_uri; *i && *i != '/' && *i != ':' && *i != '#'
+        && *i != '?'; i++) {
+        if (!_is_unreserved(*i) && !_is_pct_encoded(*i) &&
+            !_is_sub_delimin(*i)) {
+            return -EINVAL;
+        }
+    }
+
+    s.len = i - partial_uri;
+
+    *next = i;
+    SOL_DBG("Uri server: %.*s", SOL_STR_SLICE_PRINT(s));
+
+    if (server)
+        *server = s;
+
+    return 0;
+}
+
+static int
+_get_port(char *partial_uri, int *port, char **next)
+{
+    struct sol_str_slice port_slice = sol_str_slice_from_str("80");
+    char *i;
+    int r = 0;
+
+    //The server port is not required.
+    if (!*partial_uri || *partial_uri != ':')
+        goto exit;
+
+    port_slice.data = ++partial_uri;
+    for (i = partial_uri; *i && *i != '/' && *i != '?' && *i != '#'; i++) {
+        if (!isdigit(*i)) {
+            SOL_ERR("Found a non digit while parsing the port. Found: %c", *i);
+            return -EINVAL;
+        }
+    }
+    port_slice.len = i - port_slice.data;
+    *next = i;
+
+exit:
+    SOL_DBG("URI Port: %.*s", SOL_STR_SLICE_PRINT(port_slice));
+    if (port) {
+        *port = sol_util_strtol(port_slice.data, NULL, port_slice.len, 10);
+    }
+    return r;
+}
+
+static int
+_get_path(char *partial_uri, struct sol_str_slice *path, char **next)
+{
+    struct sol_str_slice p = SOL_STR_SLICE_EMPTY;
+    char *i = partial_uri;
+
+    if (!*i)
+        goto exit;
+
+    if (*i != '/')
+        goto exit;
+
+    i++;
+
+    //Oops, this is not a path.
+    if (*i == '?')
```

Comment:
I don't think this comment is correct and this special case shouldn't be handled. "/" is a valid path.